keps/0013-pull-based-audit-events-stream.md

```diff
@@ -0,0 +1,199 @@
+---
+kep-number: 13
+title: pull based audit events stream
+authors:
+  - "@CaoShuFeng"
+owning-sig: sig-auth
+participating-sigs:
+  - sig-api-machinery
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2018-06-11
+last-updated: 2018-06-11
+status: provisional
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+
+---
+
+# Pull based audit events stream
+
+## Table of Contents
+
+* [Pull based audit events stream](#pull-based-audit-events-stream)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Make audit backends dynamicly registerable](#make-audit-backends-dynamicly-registerable)
+    * [endpoints and query parameters](#endpoints-and-query-parameters)
+    * [Bandwidth limit Configuration](#bandwidth-limit-configuration)
+    * [User Stories](#user-stories)
+      * [Story 1](#story-1)
+      * [Story 2](#story-2)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+This proposal aims at providing a pull-based stream for audit logging. This would look similar to a watch request to the API server, and provide stream of audit events. The stream does not return any previous events (i.e. no associated storage), it would only stream events that occurred after the stream was opened. Two non resource-url endpoints are installed to provide the stream. '/audits/' for all audit events generated in the API server, and '/audits/{namespace}' for audits events in the specified namespace. And query parameters could be used to get subset of audit events.
+
+## Motivation
+The current audit backends (log & webhook) require very high privileges (file access to the master) or prior configuration (webhook). There are a number of use cases for dynamic and unprivileged access to the audit logs. Examples include:
```

Comment:
Most of these are addressed by [dynamic audit configuration](https://github.com/kubernetes/community/pull/2188). I think it's worth highlighting the different use cases for each here. In particular, I see the streaming endpoint as useful for short-term debugging usecases, e.g. where I run a client-side tool to pull events for a short period, and the push model is better for long-running usecases, but requires more setup.